Python: Add CuaAgentMiddleware for Computer-Use tool

f-trycua opened the pull request:
### Motivation and Context

This PR implements the integration between Microsoft Agent Framework and [Cua](https://github.com/trycua/cua) as discussed in issue #1095.

**Why is this needed?**
- Provides Agent Framework with **100+ model configurations** (OpenAI, Anthropic, OpenCUA, InternVL, UI-Tars, GLM, etc.) without duplicating model-specific parsers
- Enables **desktop automation capabilities** across Windows, macOS, and Linux through Cua's virtualization infrastructure
- Supports **composite agents** (e.g., "UI-Tars+GPT-4o") combining grounding and planning models
- Leverages Cua's existing computer-use infrastructure instead of reimplementing it

**Implementation approach:**
Following @eavanvalkenburg's guidance in #1095, this uses the `ChatMiddleware` pattern rather than implementing Cua as a Tool. This delegates the entire agent loop to Cua while maintaining Agent Framework's orchestration and human-in-the-loop capabilities.

**Why wrap `ComputerAgent` instead of just `Computer`?**
- `ComputerAgent` provides the complete agent loop (model inference → parsing → computer actions → multi-step execution) with support for 100+ model configurations
- `Computer` is just the low-level tool for executing actions (click, type, screenshot, etc.)
- By wrapping `ComputerAgent`, we get all of Cua's model support for free without reimplementing provider-agnostic parsers for OpenCUA, InternVL, UI-Tars, GLM, etc.
- This architectural choice means Agent Framework benefits from Cua's ongoing model additions automatically

**Related issue:** #1095

### Description

This PR adds `agent-framework-cua`, a new integration package that provides `CuaAgentMiddleware`.

**Key components:**

1. **`CuaAgentMiddleware`** - Middleware that intercepts chat requests and delegates to Cua's `ComputerAgent`
   - **Completely bypasses the Agent Framework chat client** by setting `context.terminate = True`
   - All model inference is handled by Cua's `ComputerAgent` (supports 100+ models)
   - Handles message format conversion between Agent Framework and Cua
   - Supports human-in-the-loop approval workflows (`require_approval`, `approval_interval`)
   - Transforms Cua results back to Agent Framework `ChatResponse` format

2. **Type definitions** - `CuaModelId`, `CuaProviderType`, `CuaOSType`, etc. for type safety

3. **Examples:**
   - `basic_example.py` - Claude Sonnet 4.5 with Linux Docker
   - `composite_agent_example.py` - UI-Tars + GPT-4o composite agent

4. **Package structure** - Follows existing integration patterns (`agent-framework-redis`, `agent-framework-mem0`)

**Architecture:**
```
Agent Framework → CuaAgentMiddleware → Cua ComputerAgent
                      ↓                      ↓
                 terminate=True    Model + Computer Loop
                                           ↓
                                       Results
                                           ↓
Agent Framework ← CuaAgentMiddleware ← Cua ComputerAgent
```

The chat client becomes a no-op since `CuaAgentMiddleware` terminates middleware execution and returns the response directly from Cua.

**Technical notes:**
- Requires Python ≥3.12 (due to `cua-agent` dependency)
- Uses dummy `chat_client` since middleware terminates execution before reaching it
- Fixed `ChatMessage.content` → `ChatMessage.text`/`contents` attribute usage in middleware

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [x] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** No

f-trycua committed:
Add CuaAgentMiddleware for Cua integration

f-trycua committed:
Add clarification that instructions are not used in examples

f-trycua committed:
Address PR feedback and restructure Cua integration samples

f-trycua commented:
Hey @ekzhu - I've addressed your feedback:

- **API Key Configuration** - Added section explaining setup via environment variables (`ANTHROPIC_API_KEY`, `OPENAI_API_KEY`, etc.)

- **Simplified Exports** - Good catch! Removed unused types from `__all__`, now only exports `CuaAgentMiddleware`

- **Restructured Samples** - Moved to `samples/getting_started/cua/` and added **workflow_orchestration** showing Agent Framework orchestration + Cua execution synergy

- **Other** - Added experimental disclaimer (DevUI pattern), Cua docs links, updated to `cua-xfce` image, clarified unused parameters

Happy to chat this week if helpful!

f-trycua committed:
Add instructions parameter support to CuaAgentMiddleware

f-trycua committed:
Update to Claude Sonnet 4.5 and prioritize Docker as default platform

f-trycua committed:
Add CuaChatClient to eliminate dummy variables

f-trycua commented:
Hey @ekzhu - I've made some improvements to the API design:

**Eliminated Dummy Variables** - Created `CuaChatClient` that properly stores model and instructions configuration. No more need for dummy `OpenAIChatClient(model_id="gpt-4o-mini", api_key="dummy-not-used")` workarounds.

**Before:**
```python
# Had to use dummy client
dummy_client = OpenAIChatClient(model_id="gpt-4o-mini", api_key="dummy-not-used")
middleware = CuaAgentMiddleware(
    computer=computer,
    model="anthropic/claude-sonnet-4-5-20250929",
    instructions="You are an assistant.",
)
agent = ChatAgent(chat_client=dummy_client, middleware=[middleware])
```

**After:**
```python
# Clean API with CuaChatClient
chat_client = CuaChatClient(
    model="anthropic/claude-sonnet-4-5-20250929",
    instructions="You are an assistant.",
)
middleware = CuaAgentMiddleware(computer=computer)
agent = ChatAgent(chat_client=chat_client, middleware=[middleware])
```

**Standardized Examples** - All samples now default to Linux on Docker (cross-platform), with macOS and Windows options shown as alternatives in comments.

Let me know if there's anything else you'd like me to address!

ekzhu reviewed (commented):
I really like the new interface! It looks very good and polished. I had some issue running it locally with Docker though -- see my comments.

I am a bit concerned about the package name "cua" being overly broad here. I think it may make sense to rename it to "trycua" or something more specific. Right now, it feels like this is the official computer-use feature of the framework. 

Another alternative is to move this package to a module inside `agent-framework-lab`, we can use the extra `cua` there.

YeIIcw committed:
Update _chat_client.py

YeIIcw committed:
Update CUA samples for compatible Docker image

Pin all CUA Docker samples to trycua/cua-xfce:latest for Windows/x64 support

Drop Anthropic instructions field so Claude requests keep working

YeIIcw committed:
Add platform setup guides for CUA samples

add Windows, macOS, and Linux quickstarts under samples/getting_started/cua/setup/

refresh the CUA README to link to the new guides and modernize prerequisites

YeIIcw commented on python/samples/getting_started/cua/basic_example/main.py:
```diff
@@ -0,0 +1,48 @@
+        # Create Cua chat client with model and instructions
+        chat_client = CuaChatClient(
+            model="anthropic/claude-sonnet-4-5-20250929",
+            instructions="You are a desktop automation assistant. Be precise and careful.",
```
> For the examples, let's set the instructions through `ChatAgent` instead. This is to keep it consistent with the rest of the samples in the repo.

Thanks for flagging this, @ekzhu! `CuaAgentMiddleware` intercepts the call and drives the run loop, so the chat client never gets a chance to apply its own system message—anything we put there gets ignored. If you need custom guidance, the simplest path is to include it in the prompt you send to `agent.run(...)`; that text is preserved and reaches CUA exactly as written.

TaoChenOSU commented on python/packages/cua/agent_framework_cua/_chat_client.py:
```diff
@@ -0,0 +1,110 @@
+                # Create middleware
+                middleware = CuaAgentMiddleware(computer=computer)
+
+                # Create agent - no dummy variables needed!
```
Question: what does it mean by "no dummy variables needed"?

YeIIcw committed:
Clean up CUA integration comments

f-trycua committed:
Merge branch 'main' into feature/cua-integration

YeIIcw committed:
Merge branch 'main' into feature/cua-integration

YeIIcw committed:
Merge branch 'main' into feature/cua-integration

YeIIcw committed:
Merge branch 'main' into feature/cua-integration

YeIIcw committed:
Merge branch 'main' into feature/cua-integration

markwallace-microsoft committed:
Merge branch 'main' into feature/cua-integration

eavanvalkenburg reviewed (commented):
Sorry it took a while to have another good look at this, but I have some structural issues with it. The most important is that this setup with the specific `CuaChatClient` doesn't work for me. The problem I have with it, is that it breaks the expectation that we want users to be able to interchange chat clients without effort. And that includes features like other middlewares on chat, local models, etc. All of that is not possible now, with this. So I think what we should do is: 1) design a computer use content type (that can be used by all major computer use capable api's like Anthropic, OpenAI and google) and 2) use that with the CuaMiddleware here, where the middleware does nothing more then look at the response, if it is computer use content, execute that and ask for another completion, if it isn't then pass the response back up the stack, and be done. That way the chat clients in AF will still be used, and we can inject the CUA Middleware without losing other functionality. let me know if that makes sense @f-trycua

eavanvalkenburg commented:
@YeIIcw that is what I envisioned initially as well, so good to see that! I am working on those content types already, as we want to ensure parity with the dotnet version and be provider agnostic, so I'll let you know when we have the design ready.

eavanvalkenburg commented:
I laid out the design here: https://github.com/microsoft/agent-framework/pull/1108 we will need to finalize it, but that should give you some idea of how we are thinking about this, please have a look and provide some feedback on what else we should have in those tools and types

eavanvalkenburg commented:
Closing, because this hasn't moved since december.
